src/System.Data.SqlClient/src/Interop/LocaleInterop.cs

```diff
@@ -0,0 +1,130 @@
+        private static extern int GetLocaleInfoEx(string lpLocaleName, uint LCType, [Out] out uint lpLCData, int cchData);
+
+        [DllImport(ApiWinCoreLocalizationObsolete, SetLastError = true, CharSet = CharSet.Unicode)]
+        private static extern int LCIDToLocaleName(uint Locale, StringBuilder lpName, int cchName, int dwFlags);
+
+        [DllImport(ApiWinCoreLocalization, SetLastError = true, CharSet = CharSet.Unicode)]
+        private static extern uint LocaleNameToLCID(string lpName, uint dwFlags);
+
+        static LocaleInterop()
+        {
+            Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
+        }
+
+        private static string LcidToLocaleNameInternal(int lcid)
+        {
+            StringBuilder localName = new StringBuilder(LOCALE_NAME_MAX_LENGTH);
+
+            if (LCIDToLocaleName(unchecked((uint)lcid), localName, localName.Capacity, 0) != 0)
+            {
+                return localName.ToString();
+            }
+            else
+            {
+                // LCIDToLocaleName should not return any other errors if we have sufficient buffer space
+                int win32Error = Marshal.GetLastWin32Error();
+                Debug.Assert(win32Error == ERROR_INVALID_PARAMETER, string.Format("Unknown error returned by LCIDToLocaleName: {0}. Lcid: {1}", win32Error, lcid));
+
+                throw new CultureNotFoundException("lcid", lcid.ToString(), message: null);
+            }
+        }
+
+        private static int LocaleNameToAnsiCodePage(string localeName)
+        {
+            uint ansiCodePage;
+            if (GetLocaleInfoEx(localeName, LOCALE_RETURN_NUMBER | LOCALE_IDEFAULTANSICODEPAGE, out ansiCodePage, sizeof(uint)) != 0)
+            {
+                return unchecked((int)ansiCodePage);
+            }
+            else
+            {
+                // GetLocaleInfoEx should not return any other errors if we have sufficient buffer space
+                int win32Error = Marshal.GetLastWin32Error();
+                Debug.Assert(win32Error == ERROR_INVALID_PARAMETER, string.Format("Unknown error returned by GetLocaleInfoEx: {0}. LocaleName: {1}", win32Error, localeName));
+
+                throw new CultureNotFoundException("localeName", localeName, message: null);
+            }
+        }
+
+        private static Tuple<string, int, Encoding> GetDetailsInternal(int lcid)
+        {
+            string localeName = LcidToLocaleNameInternal(lcid);
+            int ansiCodePage = LocaleNameToAnsiCodePage(localeName);
+            return Tuple.Create(localeName, ansiCodePage, Encoding.GetEncoding(ansiCodePage));
+        }
+
+        private static Tuple<string, int, Encoding> GetDetailsForLcid(int lcid)
+        {
+            if (lcid < 0)
+            {
+                throw ADP.ArgumentOutOfRange("lcid");
+            }
+            return s_cachedEncodings.GetOrAdd(lcid, GetDetailsInternal);
```

Comment:
Due to the C# compiler's current inability to cache delegates for delegate inference from method groups, this is going to allocate a new delegate on every call.  While counter-intuitive, it'd be better to do:
```C#
return s_cachedEncodings.GetOrAdd(lcid, lcid => GetDetailsInternal(lcid));
```